### ISSUE TITLE
Consider implementing the `int | float` special case via unsound assignability, rather than a union

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#3241]
### What are the costs of interpreting `float` as `int | float`?
1. It's a regular point of user confusion. It's much more intuitive and straightforward if a type annotation of `float` means `float` (even if that's kind of a lie at runtime -- most users will never be affected by that lie, whereas most users do encounter `int | float`).
2. It's a recurring source of bugs. At least half of the overall code/complexity (and 100% of the reported bugs) in our `NewType` implementation are due to the effective requirement to support `NewType` of a union, just so that `NewType(float)` works (even though `NewType` of a union is in general not supported.)
3. It's also a recurring source of bug reports outside of `NewType`, because it means that many common operations involving floats become stress tests of union handling in our generics solver or other areas, in a way that they aren't for most other type checkers. (This could be spun as a good thing, in that it surfaces issues, and it might eventually go away as we solidify our union handling in all these areas -- but there's no doubt that today it causes us to fail at many seemingly "simple" things involving floats that we'd otherwise handle fine.)

### What are the benefits that we get in exchange for these costs?
1. Users get sound handling of calls to e.g. `.hex()`. (I think in practice this matters to almost no-one.)
2. Users who need to pass floats to external libraries that really can only accept floats, not ints, can get diagnostics if an `int` might slip through. (I don't think this benefit exists in practice today, as it relies on these external libraries actually providing type stubs annotated to expect only-a-float, and there is currently no way to annotate that without using ty-specific extensions. So this is not a practical benefit today, it's a theoretical future benefit only if the ecosystem adopts the `int | float` understanding, and a way to annotate just-a-float.)
3. We get control-flow handling around `isinstance(x, float)` checks that understands that `x`, even if annotated as `float`, might be an `int`.

### What could we do instead?
* `int` could mean `int`, and `float` could mean `float`, with an unsound assignability carve-out to allow `int` to be assignable to `float`.
* We could add an option to turn off the unsound assignability carve-out. When this option is enabled, we could choose to continue treating `float` parameter annotations in typeshed (and maybe other third-party libraries too, with an option to override per-library) as `int | float`, until/unless typeshed starts explicitly annotating `int | float`. This allows a gradual transition (if the ecosystem ever does make this transition) from typeshed and libraries annotating `float` where they mean `int | float` to annotating `int | float` where they mean `int | float`.
* We could additionally do some things similar to what pyright does, where we distinguish "`float` that might be an `int`" (without treating it as a full-fledged union type) from "definitely just a `float`", in order to improve our control-flow handling, and retain the possibility that (given a spelling for it) some APIs could enforce taking only floats.

It doesn't seem clear to me that the benefits of the `int | float` interpretation outweigh the costs (in general, for users of Python), and given the (lukewarm at best) reception from the community so far to the idea of specifying the `int | float` interpretation, I think the cost to ty of being in the forefront of pushing this new interpretation is quite high. I don't think it's the right battle for us to invest our time in.

**Comment by sharkdp** [comment]
Thank you for writing this down. I agree with the conclusion.

There is one minor thing that I would personally add to the "costs" part of this. It's an extension of your "regular point of user confusion" item. I would argue that having `float` be interpreted as `float | int` is not just a point of confusion for *users*, it's also an annoying mental overhead for ty developers. Especially the (seemingly) self-referential part of this makes it confusing to even properly talk about this. We should probably really say that we interpret a `float` annotation as an `int | JustFloat` annotation, but then we display this as `int | float` for the benefit of users(?) who don't know what `JustFloat` is. And at that point we overloaded the term `float` which makes it hard to follow some of these discussions around it.

There might be solutions to this problem (e.g. display one of these `float`s as `float*` or similar), but I don't think that changes anything with respect to your conclusion on this topic.

**Comment by AlexWaygood** [comment]
I agree that our current approach is very confusing for both our users and us as developers, has led to a lot of bug whackamole for us, and is different to how all other type checkers approach this problem. It would be great if we could figure out a way to do better here.

Implementing this as an unsound assignability special case will probably mean we'll have to implement special cases elsewhere in our model, the same as mypy does. I think you gesture at this in your third bullet point of "What could we do instead?" when you talk about pyright's approach, but it's worth spelling it out. For example, if `float` no longer means `int | float` in this annotation, then we'll end up revealing `Never` on the final line here, because `int` and `float` are disjoint types:

```py
def f(x: float):
    if isinstance(x, int):
        reveal_type(x)
```

So we'll probably have to implement a special case such that the `float & int` intersection here doesn't solve to `Never` -- and in fact, rather than leaving it unsimplified as `float & int`, we probably want it to collapse to `int`, since it is of course impossible to have a class that is both a subclass of `float` and `int` simultaneously.

> It's a recurring source of bugs. At least half of the overall code/complexity (and 100% of the reported bugs) in our `NewType` implementation are due to the effective requirement to support `NewType` of a union, just so that `NewType(float)` works (even though `NewType` of a union is in general not supported.)

This is true -- but I suspect that reworking the `int | float` special case will lead to new, different kinds of bugs regarding `NewType`s of `float`, similar to the issue with `isinstance()` narrowing mentioned above. I.e.

```py
from typing import NewType

NT = NewType("NT", float)

def f(x: NT):
    if isinstance(x, int):
        reveal_type(x)  # ??
```

And I'm sure there will be other issues too with `NewType`s of `float`, even after making the proposed change...

Getting rid of the `int | float` special case also won't get rid of this whole genre of bugs. Recently we've seen that the awkwardness of `float` NewTypes often means that NewTypes just need to take the same code path in our implementation as other types that can wrap unions internally, such as TypeVars with unions as their upper bounds, or type aliases to unions. These other "wrapped union" pain points will persist even if we modify the way we implement the `int | float` special case.

Overall I think I'm still in favour of the change you're proposing. But it won't be easy to implement; there's no magic bullet here, unfortunately.

**Comment by sharkdp** [comment]
We discussed this internally.

Here are a few additional points that we considered:
* The type `JustFloat` isn't really all that useful at the moment, since methods on the class `float` in typeshed are annotated as returning `float`, and so that gets interpreted as `int | float` as well, e.g.:
    ```py
	def f(x: JustFloat, y: JustFloat):
	    reveal_type(x + y)  # int | float, instead of JustFloat
	```
* There's currently no immediate open problem with our behavior except for the user confusion.
* Actually implementing the special case using a carve-out in assignability sounds like it could *also* result in a lot of edge cases and bugs.

Given the second and third point here, we decided to keep our current behavior (for now).

To address the user confusion issue, we plan to change our display implementation such that `int | float` would be shown as `float` if it originates from a `float` annotation. I wrote more about this [here](https://github.com/astral-sh/ty/issues/2184#issuecomment-4343474979), and we'll use that as a tracking issue for that change.

Since that leaves nothing actionable in the issue here, I'll close this for now — the writeup here will still serve as a very useful reference in the future. Also, please feel free to comment on this issue if you feel strongly that we should reconsider this decision.

**Comment by carljm** [comment]
> * There's currently no immediate open problem with our behavior except for the user confusion.

I don't think this is true -- there are several currently open issues with `int | float` that boil down to issues with unions in generics solving. I think this is more aspirationally true: there aren't open problems here (besides user confusion) that wouldn't _in principle be fixed by_ fixing generics solving for unions in general.

**Comment by carljm** [comment]
I think #4053 is an example of an open issue where the user's desired behavior (and the behavior of all other type checkers) is fundamentally incompatible with our current approach to int/float -- it's not just user confusion (though it causes that, too!), and it isn't a bug in ty generics solving of unions.
